### PR TITLE
[core] Fix the dead loop issue in CacheManager

### LINKED ISSUE #4400: [Bug] Fix the dead loop issue in CacheManager
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.

### Paimon version

Currently caffeine2 will remove the added item directly when it was deleted before which will cause dead loop in `CacheManager`

### Compute Engine

ALL

### Minimal reproduce step

Added CI

### What doesn't meet your expectations?

cache manager

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by FangYongs**
### Purpose

Linked issue: close #4400 

### Tests

Added test case CacheManagerTest

### API and Format
no

### Documentation

no

### CODE REVIEW

**Comment by FangYongs** [comment]
@JingsongLi Please have a look at this issue, THX

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java [hunk 1]
```diff
@@ -59,7 +59,7 @@ public CacheManager(MemorySize maxMemorySize) {
 
     public MemorySegment getPage(CacheKey key, CacheReader reader, CacheCallback callback) {
         CacheValue value = cache.getIfPresent(key);
-        while (value == null || value.isClosed) {
+        while (value == null) {
```
Comment (Aitozi) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java:62]: just use if ?

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java [hunk 2]
```diff
@@ -59,7 +59,7 @@ public CacheManager(MemorySize maxMemorySize) {
 
     public MemorySegment getPage(CacheKey key, CacheReader reader, CacheCallback callback) {
         CacheValue value = cache.getIfPresent(key);
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java:61]: Can here just use `cache.get(key, Function creator)`?

Comment (FangYongs) [reply paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java]: Done

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java [hunk 3]
```diff
@@ -58,16 +60,21 @@ public CacheManager(MemorySize maxMemorySize) {
...
-            try {
-                this.fileReadCount++;
-                value = new CacheValue(MemorySegment.wrap(reader.read(key)), callback);
-            } catch (IOException e) {
-                throw new RuntimeException(e);
-            }
-            cache.put(key, value);
-        }
+        CacheValue value =
+                checkNotNull(
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java:64]: remove `checkNotNull`? it looks useless.

File: paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java [hunk 4]
```diff
@@ -95,7 +101,7 @@ private static class CacheValue {
         private final MemorySegment segment;
         private final CacheCallback callback;
 
-        private boolean isClosed = false;
+        private boolean cached = false;
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java:104]: remove this field?

Comment (FangYongs) [reply paimon-common/src/main/java/org/apache/paimon/io/cache/CacheManager.java]: done

**Comment by JingsongLi** [comment]
+1

**Pull request merged**